File: domain-persistence-modeling/java-spring-boot/src/main/java/x/scratch/child/ChildRepository.java
package x.scratch.child;

import org.springframework.data.jdbc.repository.query.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.Optional;
import java.util.stream.Stream;

import static x.scratch.Persistence.workAroundArrayTypeForPostgres;

@Repository
public interface ChildRepository
        extends CrudRepository<ChildRecord, Long> {
    @Query("SELECT * FROM child WHERE natural_id = :naturalId")
    Optional<ChildRecord> findByNaturalId(
            @Param("naturalId") String naturalId);

    @Query("SELECT * FROM child WHERE parent_natural_id = :parentNaturalId")
    Stream<ChildRecord> findByParentNaturalId(
            @Param("parentNaturalId") String parentNaturalId);

    @Query("SELECT * FROM upsert_child(:naturalId, :parentNaturalId,"
            + " :value, :defaultSideValues, :sideValues, :version)")
    Optional<ChildRecord> upsert(
            @Param("naturalId") final String naturalId,
            @Param("parentNaturalId") final String parentNaturalId,
            @Param("value") final String value,
            @Param("defaultSideValues") final String defaultSideValues,
            @Param("sideValues") final String sideValues,
            @Param("version") final Integer version);

    default Optional<ChildRecord> upsert(final ChildRecord entity) {
        final var upserted = upsert(entity.getNaturalId(),
                entity.getParentNaturalId(),
                entity.getValue(),
                workAroundArrayTypeForPostgres(entity.getDefaultSideValues()),
                workAroundArrayTypeForPostgres(entity.getSideValues()),
                entity.getVersion());
        upserted.ifPresent(entity::upsertedWith);
        return upserted;
    }
}

File: domain-persistence-modeling/kotlin-spring-boot/src/main/kotlin/x/domainpersistencemodeling/child/ChildRepository.kt
package x.domainpersistencemodeling.child

import org.springframework.data.jdbc.repository.query.Query
import org.springframework.data.repository.CrudRepository
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import x.domainpersistencemodeling.workAroundArrayTypeForPostgresWrite
import java.time.OffsetDateTime
import java.util.Optional

@Repository
interface ChildRepository : CrudRepository<ChildRecord, Long> {
    @Query("""
        SELECT *
        FROM child
        WHERE natural_id = :naturalId
        """)
    fun findByNaturalId(@Param("naturalId") naturalId: String)
            : Optional<ChildRecord>

    @Query("""
        SELECT *
        FROM child
        WHERE parent_natural_id = :parentNaturalId
        """)
    fun findByParentNaturalId(
            @Param("parentNaturalId") parentNaturalId: String)
            : Iterable<ChildRecord>

    @Query("""
        SELECT *
        FROM upsert_child(:naturalId, :otherNaturalId, :parentNaturalId,
        :state, :at, :value, :sideValues, :defaultSideValues, :version)
        """)
    fun upsert(
            @Param("naturalId") naturalId: String,
            @Param("otherNaturalId") otherNaturalId: String?,
            @Param("parentNaturalId") parentNaturalId: String?,
            @Param("state") state: String,
            @Param("at") at: OffsetDateTime, // UTC
            @Param("value") value: String?,
            @Param("sideValues") sideValues: String,
            @Param("defaultSideValues") defaultSideValues: String,
            @Param("version") version: Int)
            : Optional<ChildRecord>
}

fun ChildRepository.upsert(entity: ChildRecord): Optional<ChildRecord> {
    val upserted = upsert(entity.naturalId,
            entity.otherNaturalId,
            entity.parentNaturalId,
            entity.state,
            entity.at,
            entity.value,
            entity.sideValues.workAroundArrayTypeForPostgresWrite(),
            entity.defaultSideValues.workAroundArrayTypeForPostgresWrite(),
            entity.version)
    upserted.ifPresent {
        entity.upsertedWith(it)
    }
    return upserted
}
